src/v/config/throughput_control_group.cc

```diff
@@ -51,8 +55,30 @@ constexpr const char* tp_limit_node_in = "throughput_limit_node_in_bps";
 constexpr const char* tp_limit_node_out = "throughput_limit_node_out_bps";
 } // namespace ids
 
+constexpr char selector_prefix = '+';
+constexpr std::string_view selector_empty = "empty";
+
 static const ss::sstring noname(1, '\0');
 
+struct copyable_RE2 : re2::RE2 {
+    copyable_RE2(const copyable_RE2& other)
+      : RE2(other.pattern()) {}
+    copyable_RE2& operator=(const copyable_RE2&) = delete;
+    copyable_RE2(copyable_RE2&&) noexcept = delete;
+    copyable_RE2& operator=(copyable_RE2&&) noexcept = delete;
+    explicit copyable_RE2(const std::string& s)
+      : RE2(s) {}
+};
+
+struct client_id_matcher_type {
+    struct unspecified_tag {};
+    std::variant<unspecified_tag, copyable_RE2> v;
```

Comment:
is this what std::monostate is for, or would this be better represented as `optional<RE2>`?

Reply:
heh yeh, `optional<optional<re2>>` is sus, but generally i think the outcome is better when stuff is represented as directly as possible, even if it means weird stuff like nested optionals.